---
title: "Discovery of Acute Myeloid Leukemia Biomarkers using Ensemble Machine Learning"
output: html_notebook
---

```{r setup, include=FALSE}
require(knitr)
# knitr::opts_knit$set(root.dir = '~/Documents/GitHub/RNAseq_Cancer_Biomarkers/')
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),
                      tidy=TRUE, 
                      fig.align='center', 
                      fig.height=5, 
                      fig.width=8, 
                      dpi = 600,
                      echo=FALSE,
                      eval=TRUE)
options(stringsAsFactors = FALSE)
```

# Dependencies
```{r, eval=FALSE}
# dependency libraries
library(plyr)
library(mlr)
library(magrittr)
library(ggplot2)
library(EnsDb.Hsapiens.v75)
library(glmnet)
library(ROSE)
library(knitr)
library(stringr)
library(dplyr)
library(tibble)
library(tidyr)
library(limma)
library(edgeR)
library(MLSeq)
library(DESeq2)
library(xlsx)
library(VennDiagram)

```

```{r, eval=FALSE}
# ML functions

# Lasso
glm.binom <- function(x,y,df,ref="No", train.names=NULL, test.names=NULL, 
                      standardize=FALSE, splitIntoTrain=FALSE){
  # credit: Jenny Smith
  library(glmnet)
  #df is the matrix with the response and  gene expression. Patients as rownames.
  #x is the character vector of column names for genes 
  #y is the character vector of column names for the classifier 
  #train is a chacter vector of patient IDs
  #test is a chacter vector of Patient IDs. 
  
  response <- y
  predictors <- x
  
  #check this that referece should be the first level in glmnet package
  #Set-up the x and y matrices 
  y <- factor(df[,y])
  y <- relevel(y,ref = ref)  %>% set_names(rownames(df))
  x <- as.matrix(df[,x]) #NOTE: for categorical predictors data, should use model.matrix 
  
  
  if (any(c(is.na(y), is.na(x)))) {
    print("There Are Missing Values.")
    return(list(x=x,y=y))
  }
  
  #Check the reference level of the response.
  contrast <- contrasts(y)

  if(splitIntoTrain){
    #Use validation set approach. split observations into approx. equal groups.
    set.seed(1)  
    train <- sample(c(TRUE,FALSE), nrow(x), replace = TRUE)
    test <- (!train)
  
    train.names <- rownames(df)[train]
    test.names <- rownames(df)[test]
  }


  #grid of lambda values to test.
  grid <- 10^ seq(10,-2, length=100)
    
  #training model.
  fit <- glmnet(x[train.names,], y[train.names],
                family = "binomial",
                alpha=1,
                standardize = standardize, 
                lambda = grid, 
                intercept = FALSE)

  #use cross-validation on the training model.CV only for lambda
  set.seed(2019) 
  cv.fit <- cv.glmnet(x[train.names,], y[train.names],
                  family = "binomial",
                  type.logistic="modified.Newton", 
                  standardize = standardize,
                  lambda = grid,
                  alpha=1,
                  nfolds = length(train.names), #LOOCV 
                  type.measure = "class", 
                  intercept = FALSE)

  #Select lambda min.
  lambda.min <- cv.fit$lambda.min

  #predict the classes
  pred.class <- predict(fit, newx = x[test.names,], type="class", s=lambda.min)

  #find the test error
  tab <- table(pred.class,y[test.names])
  testError <- mean(pred.class != y[test.names]) #how many predicted classes were incorrect

  #Fit the full dataset.
  final <- glmnet(x, y,family = "binomial",
                  standardize = standardize, 
                  lambda = grid,
                  alpha = 1,
                  intercept = FALSE)

  #Extract the coefficients
  coef <- predict(final, type="coefficients", s=lambda.min)
  idx <- which(coef != 0)
  nonZero <- coef[idx,]

  #Results 
  list <- list(train.names, test.names, contrast, fit, cv.fit,tab,testError, final, nonZero)
  names(list) <- c("training.set", "testing.set","contrast", "train.fit",
                   "cv.fit", "confusionMatrix","test.error", "final.model", "nonzero.coef")
  return(list)
  
}

# SVM
runSVM <- function(seed,kerneltype="linear",trainset,trainclasses,
                   testset,testclasses, weightfilt=FALSE){
  # credit : Sean Maden
  # run SVM optimization
  # Arguments
  #   * seed : set seed (int) for randomization
  #   * kerneltype : (str) type of kernel for SVM, either 'linear' or 'gaussian'
  #   * trainset : training dataset (excluding sample classes)
  #   * trainclasses : classes for training sampels (vector) with 1:1 correspondence 
  #       with trainset rows
  #   * testset : test data (data frame or matrix), excluding classes
  #   * testclasses : classes for test samples (vector), with 1:1 row:pos correspondence
  #   * weightfilt : (FALSE or numeric) top percentage weights to use in model 
  #       (if FALSE, then all weights used) 
  # Returns
  #   * rl (list) : list containing model fitted, predictions, and performacne metrics
  require(e1071); require(ROCR)
  rl <- list(); str.options <- ""
  set.seed(seed)
  ndtr <- trainset
  ndte <- testset
  ndtr.classes <- trainclasses
  ndte.classes <- testclasses
  
  # train svm model
  svm_model <- svm(as.factor(ndtr.classes)~., 
                   data=ndtr, 
                   method="C-classification", 
                   kernel=kerneltype)
  weightsvect <- ndtr.weights <- t(svm_model$coefs) %*% svm_model$SV
  if(weightfilt){
    str.options <- c(str.options,paste0("weight filt = ",weightfilt))
    # order training data on relative weights
    ndtr.weightsort <- ndtr[,rev(order(abs(ndtr.weights)))]
    # select only top proportion weights
    nweight.col = round(ncol(ndtr.weightsort)*weightfilt,0)
    ndtr.weightfilt <- ndtr.weightsort[,c(1:nweight.col)]
    str.options <- c(str.options,paste("cols_retained:",colnames(ndtr.weightfilt),collapse=";"))
    # redefine training set, rerun SVM optimization
    ndtr <- ndtr.weightfilt
    svm_model <- svm(as.factor(ndtr.classes)~., 
                     data=ndtr, 
                     method="C-classification", 
                     kernel=kerneltype)
  } else{
    str.options <- c(str.options,"no weight filt")
  }
  pred_train <- predict(svm_model, ndtr, decision.values = TRUE)
  pred_test <- predict(svm_model, ndte, decision.values = TRUE)
  # get performance metrics
  pred <- prediction(as.numeric(attr(pred_test,"decision.values")),ndte.classes)
  perf <- performance(pred,"tpr","fpr")
  ppred <- pred_test[pred_test==1]; 
  tppred <- ndte.classes[pred_test==1]
  ppred <- as.numeric(as.character(ppred))
  testprec <- length(ppred[ppred==tppred])/length(ppred) # test precision
  rposi <- ndte.classes==1
  rtpred <- ndte.classes[rposi]; 
  rppred <- pred_test[rposi]
  rppred <- as.numeric(as.character(rppred))
  testrec <- length(rppred[rppred==1])/length(rppred) # test recall
  
  # return model, pred's, and performance metrics
  rl <- list(str.options,
             svm_model,
             weightsvect,
             pred_train,
             pred_test,
             perf,
             tppred,
             testprec,
             testrec)
  names(rl) <- c("options_string",
                 "svm_model",
                 "weightsvect",
                 "predictions_train",
                 "predictions_test",
                 "performance_test",
                 "TPR_test",
                 "precision_test",
                 "recall_test"
  )
  return(rl)
  
}

```

```{r, eval=FALSE}
# utilities for data summaries and visualization

# differential gene expression
voom_DE <- function(counts.df, ref, pheno){
  # credit: Jenny Smith
  #counts.df is a dataframe with count data, with genes as rownames
  #pheno is a character vector with patient IDs as names, and the status for each in each group(eg pos,neg)
  require(edgeR)
  library(limma)
  
  #ensure correct order for both expn and counts.df
  samples <- intersect(names(pheno), colnames(counts.df))
  pheno <- pheno[samples]
  counts.df <- counts.df[,samples]
  
  
  groups <- unique(pheno)
  groups <- c(groups[groups != ref], ref) #order so that reference is second 
  pheno.f <- factor(pheno, levels=groups)

  dge <- DGEList(counts = counts.df, group = pheno.f)

  keep.dge <- rowSums(cpm(dge) >= 1) > (0.05*ncol(counts.df)) #5% of samples with CPM >= 1
  dge <- dge[keep.dge,]
  dge <- calcNormFactors(dge)

  design <- model.matrix(~0 + pheno.f, data=dge$samples)
  colnames(design) <- levels(pheno.f)
  cont.matrix <- makeContrasts(contrasts = paste(groups, collapse = "-"), levels = design)
  
  
  v.lv <- voom(dge, design, plot = FALSE)
  

  fit <- lmFit(v.lv, design)
  fit <- contrasts.fit(fit, contrasts = cont.matrix)
  fit <- eBayes(fit)
  table <- topTable(fit, number = 20000, p.value=0.05, adjust.method="BH", sort.by="P",lfc=1)
  


  list <- list(design, v.lv, fit, table)
  names(list) <- c("desingMatrix", "voomTransformation", "fit", "DEGs")
  return(list)
}

# Survival by sample groups, plot summaries
{
  # credit: Sean Maden
  ggdat <- as.data.frame(matrix(ncol=2,nrow=0))
ggdat <- rbind(ggdat,data.frame(group='young.overallsurv',survival.time=aml.cd[class.age=='young',]$Overall.Survival.Time.in.Days))
ggdat <- rbind(ggdat,data.frame(group='young.efsurv',survival.time=aml.cd[class.age=='young',]$Event.Free.Survival.Time.in.Days))
ggdat <- rbind(ggdat,data.frame(group='old.overallsurv',survival.time=aml.cd[class.age=='old',]$Overall.Survival.Time.in.Days))
ggdat <- rbind(ggdat,data.frame(group='old.efsurv',survival.time=aml.cd[class.age=='old',]$Event.Free.Survival.Time.in.Days))

ggplot(ggdat, aes(x=ggdat$survival.time, col=ggdat$group))+geom_density()+
  theme(panel.background = element_rect(fill = 'white',colour = 'black'),
        rect = element_rect(fill = 'white',colour = "white"),
        panel.grid.major = element_line(colour = 'grey75', size=0.2),
        panel.grid.minor = element_line(colour = 'white'),
        legend.position = 'right',
        legend.background = element_rect(fill = "white", 
                                         colour ="white"),
        legend.key = element_rect(fill = "white"),
        plot.title = element_text(hjust = 0.5))+
  labs(color="Group Survival") + 
  ggtitle("Survival Time by Age Classifier")
}

# Categorize DEGs
catExpnData <- function(filenames,regex, cols, header=FALSE,removeFirstLine=FALSE, sep="\t"){
  #credit: Jenny Smith
  # Purpose: Concatenate the expression data-sets downloaded from TCGA/TARGET from GDC or any patient level data
  #eg. each individual patient has a single expression-file 
  
  library(magrittr)
  options(stringsAsFactors = FALSE)
  #filenames is a character vector of all filenames. 
  #regex is a string with the pattern to extract the patient ID , eg "^.+(Kasumi|MV4)", from filenames 
  #cols is the character vector or numeric vector of the columns to select and concatenate. 
  
  extract_cols <-function(filename,cols,rmFirstLine=FALSE){
    
    if(all(rmFirstLine & header)){
      aFile <- readLines(filename)[-1] #remove first line with extra info. 
      aFile <- str_split_fixed(aFile, pattern = "\t",n = length(cols)) %>% #split into a matrix
        set_colnames(.[1,] ) %>%  #set colnames from the first line 
        .[-1, ] #remove the header row from matrix
    }else{
      aFile <- read.delim(filename, sep=sep, header=header, as.is=TRUE)
    }
    
    output <- list()
    for ( k in 1:length(cols)){
      colname <- cols[k]
      col <- aFile[,colname]
      output[[colname]] <- col
    }
    return(output)
  }
  
  combineColumns <- function(extract_cols.res,colname){
    sapply(extract_cols.res, '[[', colname)
  }
  
  
  IDs <- gsub(regex, "\\1", filenames)
  
  columns <- lapply(filenames,extract_cols,cols=cols, rmFirstLine=removeFirstLine) %>%
    set_names(IDs)
  
  catedMatrices <- lapply(cols, combineColumns, extract_cols.res=columns)  %>%
    set_names(cols)
  
  
  return(catedMatrices)
}

# Gene summary scatter plots
{
  # credit: Sean Maden
  jpeg("target-aml_gene-meanvar-diff_test-train.jpg",10,15,units="in",res=400)
par(mfrow=c(2,1))
col.deg <- rgb(0.2,0.5,0.2,0.3)
col.all <- rgb(0.7,0.1,0.2,0.3)
test.na <- is.na(test.degdiff) | is.na(test.degvar)
plot(test.degdiff[!test.na], test.degvar[!test.na], pch=16, col=col.deg,
     main = "TARGET AML Test Subset",xlab="Gene mean diff (Low - Not-low)", ylab="Gene var diff (Low - Not-low)")
test.na <- is.na(test.alldiff) | is.na(test.allvar)
points(test.alldiff[!test.na], test.allvar[!test.na], pch=1, col=col.all)
abline(h=0,col="blue");abline(v=0,col="blue")
legend("topright",legend=c("All Genes","DEGs"),pch=c(1,16),col=c(col.all, col.deg))
train.na <- is.na(train.degdiff) | is.na(train.degvar)
plot(train.degdiff[!train.na], train.degvar[!train.na], pch=16, col=col.deg,
     main = "TARGET AML Train Subset",xlab="Gene mean diff (Low - Not-low)", ylab="Gene var diff (Low - Not-low)")
train.na <- is.na(train.alldiff) | is.na(train.allvar)
points(train.alldiff[!train.na], train.allvar[!train.na], pch=1, col=col.all)
abline(h=0,col="blue");abline(v=0,col="blue")
dev.off()
}

# Volcano plot
volcano_plot <- function(fit, cut.off=4, label.offset=0.5){
  # credit : Jenny Smith
  df <- data.frame(logFC=fit$coefficients[,1],
                   pValue=fit$p.value[,1],
                   FDR=p.adjust(fit$p.value[,1], method="BH"),
                   MeanExpression=fit$Amean) %>%
      rownames_to_column("Gene") %>%
      mutate(Neg.Log10.P= -log10(pValue),
             DEGs.Groups=case_when(
                  logFC > 1.0 & pValue < 0.05 ~ "FC Greater than 2",
                  logFC < -1.0 & pValue < 0.05 ~ "FC Less than 2",
                  TRUE ~ "Not Significant FC"))

  
  #Select differentially expressed genes to highlight in the plot. 
  ToHighlight <- df[abs(df$logFC) > cut.off & df$FDR < 0.05, "Gene"] 
  idx <- which(abs(df$logFC) > cut.off & df$FDR < 0.05)
  
  vplot <- ggplot(df, aes(x=logFC, y=Neg.Log10.P)) + 
    geom_point(data = filter(df, DEGs.Groups == "Not Significant FC"), 
               mapping = aes(x=logFC, y=Neg.Log10.P, color=DEGs.Groups), alpha=0.65)  +
    
    geom_point(data= filter(df, grepl("2", DEGs.Groups)), 
               mapping = aes(x=logFC, y=Neg.Log10.P, color=DEGs.Groups)) +
    
    geom_vline(xintercept=c(-1,1)) +
    geom_hline(yintercept = -log10(0.05)) +
    
    scale_color_manual(values=c("FC Greater than 2"="red", 
                                "FC Less than 2"="blue",
                                "Not Significant FC"="lightgrey")) +
    
    theme(plot.title = element_text(hjust = 0.5, size = 20),
          panel.background = element_rect(fill="white"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(color = "black", fill=NA),
          axis.text = element_text(color = "black"),
          axis.text.x = element_text(angle = 0,hjust=0.5,vjust = 0.5, size = 26),
          axis.text.y = element_text(size = 25),
          axis.title = element_text(size = 30),
          plot.margin = margin(2,2,2,2, unit = "mm")) +
    
    geom_text(aes(x=logFC+label.offset, y=Neg.Log10.P, label=ToHighlight),size=3.5,
              data=df[idx, ])
 

  return(vplot)
  
}

```

```{r}
# define globals
sys.sep = "/"
data.dir = "data"
seobj.dir = "seobjects"

gene.counts.tablename = "TARGET_NBL_AML_RT_WT_HTSeq_Counts.csv" # to load
clinical.tablename = "AML_assay_clinical.csv" # to load
testset.prior.name = "TARGET_AML_Testing_Samples.csv" # to load
geneanno.name = "edb_filt_anno.rda" # to load
deg.tablename = "TARGET_AML_High.Std.Risk_vs_LowRisk_DEGs.csv" # to load

countsfilt.tablename = "TARGET_AML_HTSeq_Counts.csv" # to write
tmm.counts.tablename = "TARGET_NBL_AML_WT_RT_TMMCPM_log2_Norm_Counts_17k.csv" # to write

countsseset.name <- "seset_genecounts_targetaml.rda"
tmmseset.name <- "seset_genetmmfilt_targetaml.rda"
degseset.name <- "seset_degseahack_targetaml.rda"
maeobj.name <- "mae_targetaml.rda"

# load data 
clinical <- read.csv(paste0(data.dir,sys.sep,clinical.tablename), 
                     row.names = 1, stringsAsFactors = F)

counts <- read.csv(paste0(data.dir, sys.sep, gene.counts.tablename), 
                   row.names = 1, stringsAsFactors = F)
tmm <- read.csv(paste0(data.dir, sys.sep, tmm.counts.tablename), 
                   row.names = 1, stringsAsFactors = F)
degtable <- read.csv(paste0(data.dir,sys.sep,deg.tablename), 
                     row.names = 1, stringsAsFactors = F)

testsamp.prior = read.csv(paste0(data.dir,sys.sep,testset.prior.name), row.names=1,
                          stringsAsFactors = F)

```

# Methods

## Dataset
We used [TARGET](https://ocg.cancer.gov/programs/target) pediatric AML samples. We initially focused on risk group as our primary classifier of interest. Considering sample size, demographics, and other clinical variables, we combined non-low risk groups into a single category and compared these with the low risk group. This resulted in relative balance between the categories across important clinical factors. We focused on primary tissue samples only, using [published tissue codes](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes).

## Gene Expression Data
We focused on publically available RNA-seq data, generated on the Illumina HiSeq platform, obtained for TARGET AML samples from the [Genomic Data Commons](https://portal.gdc.cancer.gov/). Gene counts were obtained using STAR 2-Pass alignment. Raw gene counts were then converted to TMM log expression using the edgeR and limma packages. 

## Preprocessing Expression Data
We then pre-filtered genes by identifying those showing greatest contrasts between our classifier groups of interest (t-test, p-adj < 0.05).

## Ensemble Machine learning.
We applied the following methods from R and Python libraries as indicated:.

# Preprocessing and Summaries

```{r}
# filter clinical variables
clinical.filt = clinical[,c(1:84)]

# filter AML clinical samples on tissue type (retain primary only)
clinical.filt = clinical.filt[clinical.filt$Diagnostic.ID %in% c("03A","09A"),]
dim(clinical.filt)
# check for repeated patient ids
summary(as.data.frame(table(clinical.filt$TARGET.USI))[,2]) # max = 1, no repeated ids

# apply sample filters to expr data
cnames.counts = colnames(counts)
counts.filt = counts[,grepl(".*\\.20\\..*",cnames.counts)] # project id filt
patidfilt = substr(colnames(counts.filt),11,16) %in% substr(clinical.filt$TARGET.USI,11,16) # patient id filt
counts.filt = counts.filt[,patidfilt]
tisstypefilt = substr(colnames(counts.filt),18,20) %in% clinical.filt$Diagnostic.ID # tissue sample type filt
counts.filt = counts.filt[,tisstypefilt]
# match clinical and counts data
counts.filt = counts.filt[,order(match(substr(colnames(counts.filt),11,16),
                                 substr(clinical.filt$TARGET.USI,11,16)
                                 )
                                 )
                          ]
identical(substr(colnames(counts.filt),11,16),substr(clinical.filt$TARGET.USI,11,16)) # true
identical(substr(colnames(counts.filt),18,20),clinical.filt$Diagnostic.ID) # true

```

```{r}
# TMM normalization
require(edgeR)
require(limma)

dge <- DGEList(counts = counts.filt)
samp <- ncol(counts.filt)
keep.dge <- rowSums(cpm(dge) >= 1) >= max(2,(0.05*samp)) #5% of AML samples has cpm of at least 1 for a gene
table(keep.dge)
# FALSE  TRUE 
# 43271 17217 
dge <- dge[keep.dge,] #subset for those genes with cmp >= 1 per gene in AML samples
dge <- calcNormFactors(dge) # Do TMM normalization
dim(dge) # NOTE: obs 17,217 retained genes, vs. 21,407 retained genes prior
cpm <- cpm(dge, log = TRUE, prior.count = 1)
head(cpm[,1:5])
write.csv(cpm, 
          paste0(data.dir, sys.sep, tmm.counts.tablename), 
          row.names = TRUE)

```

```{r}
# Differential gene expression
# DE.RG <- voom_DE(counts.df = counts.filt, ref="No",pheno=pheno)

```


```{r}
# Designate test and training data subsets

# to re-designate test/training subsets
# set.seed(2019)
#clinical.filt$exptset <- ifelse(clinical.filt$TARGET.USI %in% sample(clinical.filt$TARGET.USI, 
#                                                       size = nrow(clinical.filt)*(2/3),
#                                                       replace = FALSE),
#                  "train",
#                  "test")
#table(clinical.filt$exptset)
# test train 
#   49    96

# use prior defined test/training subsets
clinical.filt$exptset.seahack <- ifelse(clinical.filt$TARGET.USI %in% testsamp.prior[,1],
                                      "test", "train")
table(clinical.filt$exptset.seahack)

```


```{r}
# compile summarized experiment object

# gene annotations and granges objects
require(EnsDb.Hsapiens.v75)
edb <- EnsDb.Hsapiens.v75 # columns(edb) to check available annotation info
# example:
# select(edb, keys="TP53", columns=colnames(edb),keytype="SYMBOL")

# head(rownames(counts.filt))
# [1] "ENSG00000000003.13" "ENSG00000000005.5"  "ENSG00000000419.11" "ENSG00000000457.12"
# [5] "ENSG00000000460.15" "ENSG00000000938.11"
# NOTE: ids are transcript ids

# simply use gene-level information for now
genes.edb <- genes(edb)
counts.genes.grdf <- as.data.frame(matrix(ncol=6,nrow=0))
colnames(counts.genes.grdf) <- c("gene.id","gene.symbol","countsdf.id","chr.seqname","start","end")
for(i in 1:nrow(counts.filt)){
  gene.info.i = as.data.frame(genes.edb[gsub("\\..*","",rownames(counts.filt)[i])])
  counts.genes.grdf <- rbind(counts.genes.grdf,data.frame(gene.id=rownames(gene.info.i)[1],
                                                          gene.symbol=gene.info.i$gene_name[1],
                                                          countsdf.id=rownames(counts.filt)[i],
                                                          chr.seqname=gene.info.i$seqnames,
                                                          start=gene.info.i$start,
                                                          end=gene.info.i$end,
                                                          stringsAsFactors = F))
  message(i," perc. complete = ",round(100*(i/nrow(counts.filt)),4),"%")
}
save(counts.genes.grdf, file=paste0(data.dir, sys.sep, geneanno.name))

# se experiments using filtered genes
length(intersect(rownames(counts.filt), counts.genes.grdf$countsdf.id)) # 54713
length(intersect(counts.genes.grdf$countsdf.id, rownames(dge))) # 16696

counts.se <- counts.filt[rownames(counts.filt) %in% counts.genes.grdf$countsdf.id,]
dge.se <- dge[rownames(dge) %in% counts.genes.grdf$countsdf.id,]

# order genes for counts se
ganno.counts <- counts.genes.grdf[order(match(counts.genes.grdf$countsdf.id,
                                              rownames(counts.se))),]
identical(counts.genes.grdf$countsdf.id, rownames(counts.se))
ganno.tmm <- counts.genes.grdf[counts.genes.grdf$countsdf.id %in% rownames(dge.se),]
ganno.tmm <- ganno.tmm[order(match(ganno.tmm$countsdf.id,rownames(dge.se))),]
identical(ganno.tmm$countsdf.id,rownames(dge.se))
colnames(ganno.counts) <- colnames(ganno.tmm) <- c("gene.id","gene.symbol","countsdf.id","seqnames","start","end")

require(GenomicRanges)
ggr.counts <- makeGRangesFromDataFrame(ganno.counts, 
                                       keep.extra.columns = T, 
                                       ignore.strand = T)
names(ggr.counts) <- ggr.counts$countsdf.id
ggr.tmm <- makeGRangesFromDataFrame(ganno.tmm,
                                    keep.extra.columns = T,
                                    ignore.strand = T)
names(ggr.tmm) <- ggr.tmm$countsdf.id

require(SummarizedExperiment)
# Gene Expr Counts SE object
identical(ggr.counts$countsdf.id, rownames(counts.se)) # TRUE
identical(names(ggr.counts), rownames(counts.se)) # TRUE
identical(substr(colnames(counts.se),11,16), 
          substr(clinical.filt$TARGET.USI,11,16)) # TRUE
counts.seset <- SummarizedExperiment(assays = as.matrix(counts.se),
                     rowRanges = ggr.counts, 
                     colData = DataFrame(clinical.filt, 
                                         row.names = colnames(counts.se)
                                         ),
                     metadata = list(dataset = "TARGET_AML", 
                                     assay_source = "GDC",
                                     genome_build = "hg19")
                     )

# Gene TMM SE object
identical(ggr.tmm$countsdf.id, rownames(dge.se)) # TRUE
identical(names(ggr.tmm), rownames(dge.se))
identical(substr(colnames(dge.se),11,16), 
          substr(clinical.filt$TARGET.USI,11,16)) # TRUE
tmm.seset <- SummarizedExperiment(assays = as.matrix(dge.se),
                                  rowRanges = ggr.tmm,
                                  colData = DataFrame(clinical.filt,
                                                      row.names = colnames(dge.se)
                                                      ),
                                  metadata = list(dataset = "TARGET_AML",
                                                  assay_source = "GDC",
                                                  genome_build = "hg19",
                                                  normalization_strategy = "TMM, log2_normcounts, limma, edgeR"))

# DEG TMM SE object
deglist = rownames(degtable)
length(intersect(deglist, counts.genes.grdf$countsdf.id)) # 1937 of 1998
degfilt = deglist[deglist %in% counts.genes.grdf$countsdf.id]
ggr.deg = ggr.counts[names(ggr.counts) %in% degfilt]
deg.assay <- counts.se[rownames(counts.se) %in% degfilt,]
ggr.deg <- ggr.deg[order(match(names(ggr.deg), rownames(deg.assay)))]
identical(names(ggr.deg), rownames(deg.assay)) # TRUE
identical(substr(colnames(dge.se),11,16), 
          substr(clinical.filt$TARGET.USI,11,16)) # TRUE
# add the deg statistics to gene annotation
degstats = degtable[rownames(degtable) %in% degfilt,]
degstats = degstats[order(match(rownames(degstats), names(ggr.deg))),]
identical(rownames(degstats), names(ggr.deg)) # TRUE
ggr.deg$logFC <- degstats$logFC
ggr.deg$AveExpr <- degstats$AveExpr
ggr.deg$t <- degstats$t
ggr.deg$p.unadj <- degstats$P.Value
ggr.deg$p.adj.bh <- degstats$adj.P.Val
ggr.deg$b <- degstats$B
# make the se object
deg.seset <- SummarizedExperiment(assays = as.matrix(deg.assay),
                                  rowRanges = ggr.deg,
                                  colData = DataFrame(clinical.filt,
                                                      row.names = colnames(deg.assay)
                                                      ),
                                  metadata = list(dataset = "TARGET_AML",
                                                  assay_source = "GDC",
                                                  genome_build = "hg19",
                                                  normalization_strategy = "DEG RiskGroup, Low=0 notLow=1, reference: Low, TMM_log2_normcounts, voom_DE function"))

# Save new SE objects
save(counts.seset, file=paste0(seobj.dir, sys.sep, countsseset.name))
save(tmm.seset, file=paste0(seobj.dir, sys.sep, tmmseset.name))
save(deg.seset, file=paste0(seobj.dir, sys.sep, degseset.name))

# Multi Assay Experiment class
require(MultiAssayExperiment)
counts.map <- data.frame(primary = colnames(counts.seset),
                         colname = colnames(counts.seset),
                         stringsAsFactors = F)
tmm.map <- data.frame(primary = colnames(tmm.seset),
                         colname = colnames(tmm.seset),
                         stringsAsFactors = F)
deg.map <- data.frame(primary = colnames(deg.seset),
                         colname = colnames(deg.seset),
                         stringsAsFactors = F)
listmap <- list(counts.map, tmm.map, deg.map)
names(listmap) <- c("gene_counts", "tmm_log2norm_filtcounts", "deg_tmm_counts")
dfmap <- listToMap(listmap)
objlist = list("gene_counts" = counts.seset,
               "tmm_log2norm_filtcounts" = tmm.seset,
               "deg_tmm_counts" = deg.seset)
patient.data <- clinical.filt
rownames(patient.data) <- colnames(counts.se)
targetaml.mae <- MultiAssayExperiment(objlist, patient.data, dfmap)
save(targetaml.mae, file=paste0(seobj.dir, sys.sep, maeobj.name))
```

## Dataset Summaries

```{r}
# demographics


```

## Gene Normalization
```{r}
```

## DEG Computation
```{r}
```

## DEG Correlation Heatmap
```{r}
```

# Results

## Summary Statistics

## Model Fitting

## Feature Selection

## Consensus Feature Validation

# Conclusions
