---
title: 'GDC data download'
author: "Jenny Smith"
date: "February 2, 2018"
output: html_document
---


#Set-up 

```{r setup}
library(knitr)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=50),tidy=TRUE, fig.align='center', fig.height=5, fig.width=8, dpi = 600)
knitr::opts_knit$set(root.dir = '~/Documents/GitHub/RNAseq_Cancer_Biomarkers/')
options(stringsAsFactors = FALSE)
```


```{r message = FALSE, warning=FALSE}
library(stringr)
library(magrittr)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
getwd()
```


#References

https://bioconductor.org/packages/release/bioc/vignettes/GenomicDataCommons/inst/doc/overview.html#quickstart
http://bioconductor.org/packages/release/bioc/vignettes/TCGAutils/inst/doc/TCGAutils.html

http://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/index.html
http://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/query.html#get_manifest_file


#Installation  

```{r}
# BiocManager::install("TCGAbiolinks") #as back up 
# library(TCGAbiolinks)
```

```{r}
# BiocManager::install('GenomicDataCommons')
library(GenomicDataCommons)
```

```{r}
GenomicDataCommons::status()
```

```{r}
stopifnot(GenomicDataCommons::status()$status=="OK")
```


#Orientation with Package 

There are 4 main types of objects:
1. projects()
2. files()
3. cases() 
4. annotations()

Each of these can be filtered and specific fields can be selected. 
Then piped (`%>%`) into `results()` or `results_all()` to get a list with each file/case ID and its metadata (like date sumbitted to GDC)


```{r}
available_fields('projects') %>% length()
available_fields('files') %>% length()
available_fields('cases') %>% length()
available_fields('annotations') %>% length()
```

##Projects 

```{r}
pQuery = projects()
head(projects())
# length(pQuery) #5 
```

```{r}
default_fields(pQuery)
```

```{r}
presults = pQuery %>% results() #results_all()
class(presults)
str(presults)
ids(presults) #10 datasets returned by default when no arguments given to projects and no filters added
```


##Files 

```{r}
default_fields(files()) #data types - eg RNA-seq, DNAme, miRNA-seq, etc. 

files() %>% count() #without filters, and only 10 Cohorts has 358,679 files
```


##Cases 

```{r}
qcases = cases() #patient samples
qcases$fields #grep_fields(), field_picker() can be useful here
head(available_fields('cases')) #clinical data elements "columns"

qcases %>% count() #33,096 patient cases/samples
```



#Summarize Available Information 


```{r}
# total number of files of a specific type. 
#remember to use available_feilds(files()) for your facet arguments. 
res = files() %>% 
  facet(c('type','data_type','data_format',
          'cases.project.project_id')) %>% #Default is to set facets for all default fields. 
  aggregations()

# res$type
# res$data_type
# res$data_format
res$cases.project.project_id$key  %>% .[order(.)]
```

```{r}
res.projects <- projects() %>%
  facet(c("project_id")) %>%
  aggregations() 

# str(res.projects)
res.projects
```

```{r}
res.analysis <- files() %>%
  facet(grep_fields('files','analysis.workflow')) %>% 
  aggregations()

# res.analysis$analysis.workflow_version
# res.analysis$analysis.workflow_type
```

```{r}
res.cases <- cases() %>% 
  facet() %>% #Default is to set facets for all default fields. 
  aggregations()

head(res.cases$primary_site)
```


#Query GDC for Gene Expression Data and Clinical Data

```{r}
grep('project',available_fields('files'),value=TRUE) #see facets above
grep('type',available_fields('files'),value=TRUE) 
grep('cases.samples.submitter_id',available_fields('files'), value=TRUE)
```

 
```{r}
#Note: Here I need the sample UUID because it will map file ID to TARGET Barcode later, and is not included in the default fields. 
#so I used some selected fields, such as project ID, associated_entities (Sample UUIDs),and workflow type, in addition to  the default.

qfiles <- files(fields=c("cases.project.project_id",default_fields('files'), grep_fields('files', "associated_entities"), 
                         "analysis.analysis_type", "analysis.workflow_type", "analysis.workflow_version")) %>% 
  filter(~ type == 'gene_expression' & 
           analysis.workflow_type == 'HTSeq - Counts' &
          (cases.project.project_id == "TARGET-AML" | cases.project.project_id == "TARGET-NBL" | cases.project.project_id == "TARGET-WT" | cases.project.project_id == "TARGET-CCSK" | cases.project.project_id == "TARGET-OS" | cases.project.project_id == "TARGET-RT")) 


qfiles %>% count() #473
# head(qfiles)
```

NOTE: OS and CCSK have no mRNAseq. TARGET-RT has only 5 patient samples with RNAseq on GDC


#Create ID Map for Manifest, Clinical, and Expression Files

```{r}
res.expn <-  qfiles %>%
  results_all()

length(res.expn)
names(res.expn)
# str(res.expn)
# lapply(res.expn,head)
```

The "cases" feilds - if used in `results()` contains much of the clinical data elements that are extracted through `gdc_clinical()` function used below. 

```{r}
#there are 2 Sample IDs associated with a single file. 
#Filter these out, since they are suspect. 
idx <- sapply(res.expn$associated_entities , nrow) %>% grep(2, .)
```

```{r}
#check that these are the expected files/samples which more than 1 ID associated

# res.expn$associated_entities[idx]
# res.expn$cases[idx]
```

```{r}
ID.map <- res.expn[ !grepl("list|data.frame", sapply(res.expn, class)) ] %>%
  as.data.frame() %>%
  slice(-idx) %>% #remove the multiple ID mapped files
  mutate(project.project_id=unlist(res.expn$cases[-idx])) %>%
  bind_cols(., bind_rows(res.expn$associated_entities[-idx])) #associated_entities was a list of data.frames, converted a single data.frame with bind_rows()

head(ID.map)
dim(ID.map) #474

# write.csv(ID.map, "GCD_TARGET_Data_Manifest_AML_NBL_WT_RT.csv", row.names = FALSE)
```

```{r}
table(ID.map$project.project_id)
```


#Create a Download Manifest file

```{r}
qfiles <- qfiles %>%
  filter(~ file_id %in% names(res.expn$associated_entities[-idx]) )

qfiles %>% count() #469
```

```{r}
manifest_df = qfiles %>% manifest()
head(manifest_df)
dim(manifest_df) #469 


# write.table(manifest_df, "TARGET_Manifest_RNAseq_Counts.txt", row.names = FALSE, sep="\t", quote=FALSE)
```
 
 

#Download the files 

```{r}
#Add the GDC Client to your R options
# options(gdc_client="/home/jlsmith3/scripts/opt/bin/gdc-client")
# gdc_client()
```

```{r}
#Set your destination directory (be aware this downloads your current working directory, unless set otherwise)
dir.create("Expn_Data")
gdc_set_cache(directory = "Expn_Data/")
```

```{r}
#download step
fnames = gdcdata(manifest_df$id,progress=FALSE,access_method = "api", use_cached = FALSE)

head(fnames)
```


#Check MD5 Sums

```{r}
#check MD5 sums
```



#Reorganize and Concatenate the Downloaded Files 

```{r}
source("Cat_Expn_Data.r")
```

NOTE: here i had to use some bash reprocessing 
1. mv all counts files to a directory called Expn_Data/ (remove the nested directories)
2. gunzip *.gz (some were gzipped, some not... )

```{r}
path="~/Documents/GitHub/RNAseq_Cancer_Biomarkers/Expn_Data/"

files <- dir(path=path, pattern="*.counts", recursive = TRUE) %>%
  paste0(path, .)

head(files)
```

```{r}
cated <- catExpnData(filenames = files, 
                     regex="^.+\\/([a-z0-9].+.htseq.counts)", 
                     cols=c(1,2), header = FALSE)
```

```{r}
lapply(cated, dim)
all(lapply(cated$`1`, function(x) identical(x, cated$`1`[,1]))) #check order is identical for all genes columns 
```

```{r}
counts <- cated$`2` %>%
  as.data.frame() %>% 
  mutate(Genes=cated$`1`[,1]) %>% #genes in same order in all datasets. 
  dplyr::select(Genes, everything())

head(counts[,1:10])


```

```{r}
ID.map.sub <- dplyr::select(ID.map, file_name, entity_submitter_id) %>% 
  mutate_at(vars(file_name, entity_submitter_id), funs(gsub("-","\\.", .)))%>%
  mutate_at(vars(file_name), funs(gsub(".gz", "",. )))

fixNames.df <- data.frame(names= colnames(counts)[-1] %>% gsub("^X", "",.))  %>%
  left_join(.,ID.map.sub, by=c("names"="file_name"))


head(fixNames.df)
dim(fixNames.df)
```

```{r}
#Finally, update the column names
colnames(counts)[-1] <- fixNames.df$entity_submitter_id

head(counts[,1:10])
```


```{r}
# write.csv(counts, "TARGET_NBL_AML_RT_WT_HTSeq_Counts.csv", row.names = FALSE)
```





#Download Clinical Data 

https://seandavi.github.io/2018/03/extracting-clinical-information-using-the-genomicdatacommons-package/


##from GDC Database

```{r}
cases() %>% filter(~ project.project_id == "TARGET-AML") %>% count() #988 patients

case_ids <- cases() %>%
  filter(~ project.project_id == "TARGET-AML") %>%
  ids()

head(case_ids)
```

```{r}
#download the clinical data per patient
clin_res = gdc_clinical(case_ids)

sapply(clin_res, dim) %>% 
  set_rownames(c("rows","colunms"))

# sapply(clin_res, colnames)
# sapply(clin_res, head)
```

```{r}
#Combine all Clinical Data into a Master File
full_clin = with(clin_res,
     main %>% #what is main here? not a function...
     left_join(demographic, by = "case_id") %>%
     left_join(exposures, by = "case_id") %>%
     left_join(diagnoses, by = "case_id"))

head(full_clin)
dim(full_clin) # 988  33
```

##Precompiled Clinical Data Elements from Study Authors 

```{r}
qClin <- files() %>%
    filter(~ type == 'clinical_supplement' &
             (cases.project.project_id == "TARGET-AML"| 
             cases.project.project_id == "TARGET-NBL" | 
             cases.project.project_id == "TARGET-WT"  |
           cases.project.project_id == "TARGET-RT")) 

# str(qClin)
qClin %>% count()
```


```{r}
manifest.clin <- manifest(qClin)
head(manifest.clin)
```


```{r}
gdc_set_cache(directory = "~/Documents/GitHub/RNAseq_Cancer_Biomarkers")

clinFile <- gdcdata(uuids = manifest.clin$id)
clinFile
```


The downloaded data is each in thier own directories, with the file UUID as the directory name.
Need to organize and move the Clinical Data so it is more easily usable. 

```{r}
system("ls -1")
system("pwd")


#didn't work--- need to figure out how . Works in terminal directly
system("find . -type f -name '*.xlsx' -exec mv {} . \;")

#note this below works since there were NOT other directories except the file UUID dirs.
#Do not use this if you have others present (though non-empty directories will just result in an error)
system("for dir in $(ls -1d */); do rmdir $dir; done")
```






#Session Information 

```{r}
sessionInfo()
```

